docs/decisions/0031-automatic-tool-calling-controls.md

```diff
@@ -0,0 +1,198 @@
+
+    // ...
+}
+```
+
+The planner could then use this metadata to determine how many model iterations occurred as a result of each request, and enforce the `MaxIterations` limit accordingly. This would also reduce the overall model calls made by the planner, since tool calls could be "chained" consecutively by the kernel rather than having a chat response in between each one (as would be the case with Option 1).
+
+Updated helper method in _FunctionCallingStepwisePlanner.cs_:
+
+```csharp
+private async Task<ChatMessageContent> GetCompletionWithFunctionsAsync(
+    int iterationsCompleted,
+    ChatHistory chatHistory,
+    Kernel kernel,
+    IChatCompletionService chatCompletion,
+    OpenAIPromptExecutionSettings openAIExecutionSettings,
+    ILogger logger,
+    CancellationToken cancellationToken)
+{
+    openAIExecutionSettings.ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions;
+    openAIExecutionSettings.ToolCallBehavior.MaximumUseAttempts = this.Config.MaxIterations - iterationsCompleted; // limit the number of tool calls to the number of iterations remaining
+
+    await this.ValidateTokenCountAsync(chatHistory, kernel, logger, openAIExecutionSettings, cancellationToken).ConfigureAwait(false);
+    return await chatCompletion.GetChatMessageContentAsync(chatHistory, openAIExecutionSettings, kernel, cancellationToken).ConfigureAwait(false);
+}
+```
+
+Updated loop in _FunctionCallingStepwisePlanner.cs_:
+
+```csharp
+for (int iteration = 0; iteration < this.Config.MaxIterations; /* iteration is incremented within the loop */)
+{
+    // sleep for a bit to avoid rate limiting
+    // ...
+
+    // For each step, request another completion to select a function for that step
+    chatHistoryForSteps.AddUserMessage(StepwiseUserMessage);
+    var chatResult = await this.GetCompletionWithFunctionsAsync(iteration, chatHistoryForSteps, clonedKernel, chatCompletion, stepExecutionSettings, logger, cancellationToken).ConfigureAwait(false);
+    chatHistoryForSteps.Add(chatResult);
+
+    // Increment iteration based on the number of model round trips that occurred as a result of the request
+    object? value = null;
+    chatResult.Metadata?.TryGetValue("Iterations", out value);
+    if (value is not null and int)
+    {
+        iteration += (int)value;
+    }
+    else
+    {
+        // Could not find iterations in metadata, so assume just one
+        iteration++;
+    }
+
+    // Check for final answer
+    // ...
+}
+```
+
+### Option #3: Add a pre-invoke callback and track model iterations
+
+Add a PreInvokeCallback field to `ToolCallBehavior`. This callback would be invoked inside `ClientCore.cs` before a tool call is invoked. If `true`, would proceed with the tool call invocation. This approach would give the caller control over whether or not a tool call is invoked, while still leveraging the kernel's code for invocation.
```

Comment:
I think some variation of option 3 is the most robust / future-proof route. I imagine we'd want something like (waving my hands):
```C#
class ToolCallBehavior
{
    public static ToolCallBehavior Create(Func<AutoInvokeKernelSettings, ValueTask<AutoInvokeKernelResult>> callback);
}
```
where `AutoInvokeKernelSettings` and `AutoInvokeKernelResult` are classes with various information on them, with the input providing details about the Kernel/KernelFunction/KernelArguments/NumberOfIterations/etc., and the output storing details that inform the processing loop of the FunctionResult, what to do next, etc.  Just a rough sketch.